### ISSUE TITLE
torch.median slower than torch.sort

### ISSUE THREAD

**Issue by mberr** [pytorch/pytorch#51450]
## 🐛 Bug

Using `torch.median(x)` is slower than explicitly sorting the array with `torch.sort()` and selecting the middle element when the tensor resides on cpu.

## To Reproduce

```python-console
>>> from torch.utils.benchmark import Timer
>>> t = Timer(stmt="torch.median(x, dim=0)", setup="x=torch.rand(11, 1000, 4000, device='cpu')")
<torch.utils.benchmark.utils.common.Measurement object at 0x7f8e5527ba60>
torch.median(x, dim=0)
  19.04 s
  1 measurement, 1 runs , 1 thread
>>> t = Timer(stmt="torch.sort(x, dim=0).values[5]", setup="x=torch.rand(11, 1000, 4000, device='cpu')")
<torch.utils.benchmark.utils.common.Measurement object at 0x7f8e584a7880>
torch.sort(x, dim=0).values[5]
  856.96 ms
  1 measurement, 1 runs , 1 thread
```

For CUDA tensors I did not observe this large difference.
```python-console
>>> from torch.utils.benchmark import Timer
>>> t = Timer(stmt="torch.median(x, dim=0)", setup="x=torch.rand(11, 1000, 4000, device='cuda')")
<torch.utils.benchmark.utils.common.Measurement object at 0x7f85036f86d0>
torch.median(x, dim=0)
  56.26 ms
  1 measurement, 10 runs , 1 thread
>>> t = Timer(stmt="torch.sort(x, dim=0).values[5]", setup="x=torch.rand(11, 1000, 4000, device='cuda')")
<torch.utils.benchmark.utils.common.Measurement object at 0x7f85036192b0>
torch.sort(x, dim=0).values[5]
  39.60 ms
  1 measurement, 10 runs , 1 thread
```

## Expected behavior

Since the median operation does not *require* to sort the whole tensor, it should be at least as fast the sort operation.

## Environment

Collecting environment information...
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 450.102.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

**Comment by mruberry** [comment]
Thank you for reporting this issue, @mberr. It'd be nice to have a faster median!

**Comment by pkienzle** [comment]
Here's what I'm using:
```
def amedian(x, dim, keepdim=False):
    """
    Find the median along a particular dimension.

    If the dimension length is even take the average of the central values.

    Use *keepdim=True* to preserve the dimension after reduction.
    """
    index = torch.argsort(x, dim)
    deref = [slice(None, None)]*len(x.shape)
    middle = x.shape[dim]//2
    even = 1 - x.shape[dim]%2
    deref[dim] = slice(middle-even, middle+1+even)
    values = x.gather(dim, index[deref])
    return (
        values.mean(dim, keepdim=keepdim) if even 
        else values if keepdim 
        else values.squeeze(dim))
```

Time for `[7, 3, 3520, 4656]` along dimension 1 on 2080Ti

    amedian(x, 1): 0.525 s
    torch.median(x, 1)[0]: 110.7 s

**Comment by min-jean-cho** [comment]
@mberr, I could not reproduce the performance gap with torch 1.13.0+cpu. You may want to have a try with the latest torch version. 

Tested on Xeon 8180, 28 cores * 2 sockets with HT,
|        size       	| torch.median(x, dim=0) 	| torch.sort(x,dim=0).values[len(x)//2] 	|
|:-----------------:	|------------------------	|---------------------------------------	|
| (11, 1000, 4000)  	| 866.83 ms              	| 952.02 ms                             	|
| (3, 1000, 1000)   	| 42.28 ms               	| 51.37 ms                              	|
| (9, 1000, 1000)   	| 166.89 ms              	| 185.03 ms                             	|
| (33, 1000, 1000)  	| 637.99 ms              	| 1.09 s                                	|
| (65, 1000, 1000)  	| 1.13 s                 	| 2.51 s                                	|
| (129, 1000, 1000) 	| 2.13 s                 	| 5.84 s                                	|

**Comment by mberr** [comment]
Hi @min-jean-cho , it has been quite a while, and I do not have access to the same machine anymore. However, I re-ran the script (with some minor modifications due to changed functionality in `Timer`) on a different machine, and I could not observe a big difference between the two. So likely this has been fixed somewhere between torch 1.7 and 1.13 :sweat_smile: 

```python-console
>>> from torch.utils.benchmark import Timer
>>> Timer(stmt="torch.median(x, dim=0)", setup="x=torch.rand(11, 1000, 4000, device='cpu')").blocked_autorange()
<torch.utils.benchmark.utils.common.Measurement object at 0x7f3d8ae45240>
torch.median(x, dim=0)
setup: x=torch.rand(11, 1000, 4000, device='cpu')
  999.05 ms
  1 measurement, 1 runs , 1 thread

>>> Timer(stmt="torch.sort(x, dim=0).values[5]", setup="x=torch.rand(11, 1000, 4000, device='cpu')").blocked_autorange()
<torch.utils.benchmark.utils.common.Measurement object at 0x7f3d8ae24490>
torch.sort(x, dim=0).values[5]
setup: x=torch.rand(11, 1000, 4000, device='cpu')
  1.16 s
  1 measurement, 1 runs , 1 thread

```

<summary>Environment</summary>
<details>

```
Collecting environment information...
PyTorch version: 1.13.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.10 (x86_64)
GCC version: (Ubuntu 12.2.0-3ubuntu1) 12.2.0
Clang version: Could not collect
CMake version: version 3.24.2
Libc version: glibc-2.36

Python version: 3.10.7 (main, Nov  2 2022, 18:49:29) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.19.0-23-generic-x86_64-with-glibc2.36
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce 930MX
Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.1
[pip3] pytorch-lightning==1.7.6
[pip3] torch==1.13.0+cpu
[pip3] torch-max-mem==0.0.4
[pip3] torch-ppr==0.0.8
[pip3] torchaudio==0.13.0+cpu
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.14.0+cpu
[conda] Could not collect
```
</details>
